src/octree/batch_iterator.rs

```diff
@@ -144,29 +154,46 @@ impl<'a> BatchIterator<'a> {
     }
 
     /// compute a function while iterating on a batch of points
-    pub fn try_for_each_batch<F>(&mut self, mut func: F) -> Result<()>
+    pub fn try_for_each_batch<F>(&mut self, func: F) -> Result<()>
     where
         F: FnMut(PointData) -> Result<()>,
     {
-        //TODO(catevita): mutable function parallelization
+        let octree = self.octree;
+        // nodes iterator: retrieve nodes
+        let node_id_iterator = octree.nodes_in_location(self.point_location);
+
         let local_from_global = self
             .point_location
             .global_from_local
             .clone()
             .map(|t| t.inverse());
-        let mut point_stream = PointStream::new(self.batch_size, local_from_global, &mut func);
-        // nodes iterator: retrieve nodes
-        let node_id_iterator = self.octree.nodes_in_location(self.point_location);
-        // operate on nodes
-        for node_id in node_id_iterator {
-            let point_iterator = self.octree.points_in_node(self.point_location, node_id);
-            for point in point_iterator {
-                point_stream.push_point_and_callback(point)?;
+        // operate on nodes: one thread for each node
+        crossbeam::scope(|s| {
+            let (tx, rx) = crossbeam::channel::bounded::<PointData>(2);
+            for node_id in node_id_iterator {
+                let tx = tx.clone();
+                let local_from_global = local_from_global.clone();
+                let point_location = self.point_location.clone();
+                let batch_size = self.batch_size;
+
+                s.spawn(move |_| {
+                    let send_func = |batch: PointData| match tx.send(batch) {
+                        Ok(_) => Ok(()),
+                        Err(e) => Err(ErrorKind::Channel(format!(
+                            "sending operation failed, nothing more to do {:?}",
+                            e,
+                        ))
+                        .into()),
+                    };
+                    let point_iterator = octree.points_in_node(&point_location, node_id);
+                    let mut point_stream =
+                        PointStream::new(batch_size, local_from_global, &send_func);
+                    let _ = point_stream.push_point_and_callback(point_iterator);
+                });
             }
-        }
-        // TODO(catevita): return point data through mpsc channel
-        // TODO(catevita): apply mut function to received data
 
-        Ok(())
+            rx.iter().try_for_each(func)
+        })
+        .expect("child thread panic")
```

Comment:
Nit: How about "Panic in try_for_each_batch child thread"?

Reply:
i added also BatchIterator in front of your sentence